### PR TITLE
RES-885: Refactor linearsdr and cnnsdr classes

### PR DESCRIPTION
**Pull request by lscheinkman**
@subutai Please review.
Here is an example on how to use the new API:
```
import torch.nn as nn
import htmresearch.frameworks.pytorch.modules as htm
:
:
# Equivalent to LinearSDR class

self.l1 = htm.SparseWeights(nn.Linear(28*28, 500), 0.4)

self.kw = htm.KWinners(n=500, k=50, kInferenceFactor=1.5,
                       boostStrength=1.0, boostStrengthFactor=0.9)
```
and in the `forward` method just call ```self.kw(self.l1(x))```

The same API is available in 2d for CNN.

### CODE REVIEW

File: htmresearch/frameworks/pytorch/functions/k_winners.py [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+
+from __future__ import print_function
+
+import torch
+
+
+
+class k_winners(torch.autograd.Function):
+  """
+  A simplistic K-winner take all autograd function for experimenting with
```
Comment (subutai) [inline htmresearch/frameworks/pytorch/functions/k_winners.py]: Maybe rephrase to "A simple K-winner take all autograd function for creating layers with sparse output."

File: htmresearch/frameworks/pytorch/functions/k_winners.py [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+    res = res.reshape(x.shape)
+
+    ctx.save_for_backward(indices)
+    return res
+
+
+  @staticmethod
+  def backward(ctx, grad_output):
+    """
+    In the backward pass we receive the context object and a Tensor containing
```
Comment (subutai) [inline htmresearch/frameworks/pytorch/functions/k_winners.py]: "In the backward pass, we set the gradient to 1 for the winning units, and 0 for the others."

File: htmresearch/frameworks/pytorch/modules/k_winners.py [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+    self.learningIterations += batchSize
+
+    scaleFactor = float(x.shape[2] * x.shape[3])
+    period = min(self.dutyCyclePeriod, self.learningIterations)
+    self.dutyCycle.mul_(period - batchSize)
+    s = x.gt(0).sum(dim=(0, 2, 3), dtype=torch.float) / scaleFactor
+    self.dutyCycle.reshape(-1).add_(s)
+    self.dutyCycle.div_(period)
+
+    # This seems to happen, but very rarely. Keep an eye on it.
```
Comment (subutai) [inline htmresearch/frameworks/pytorch/modules/k_winners.py]: Maybe get rid of this if clause?

File: htmresearch/frameworks/pytorch/sparse_net.py [hunk 4]
```diff
@@ -275,29 +320,8 @@ def __init__(self,
...
-        if hasattr(module, "setBoostStrength"):
-          module.setBoostStrength(self.boostStrength)
-        if hasattr(module, "rezeroWeights"):
-          # The optimizer is updating the weights during training after the forward
-          # step. Therefore we need to re-zero the weights after every epoch
-          module.rezeroWeights()
-
-      print("boostStrength is now:", self.boostStrength)
+    self.apply(htm.updateBoostStrength)
+    self.apply(htm.rezeroWeights)
```
Comment (subutai) [inline htmresearch/frameworks/pytorch/sparse_net.py]: Does apply call these functions recursively on every module?

Comment (lscheinkman) [reply htmresearch/frameworks/pytorch/sparse_net.py]: yes

File: htmresearch/frameworks/pytorch/functions/k_winners.py (hunk 2 above)
Comment (lscheinkman) [reply htmresearch/frameworks/pytorch/functions/k_winners.py]: I just moved the k-winners file to the `functions` folder without looking at the comments. I also forgot to delete the original file in the parent folder. 
I will update the comments and delete the copy in the parent folder. 
Should I also delete the `linear_sdr` and `cnn_sdr` files? These files are now obsolete.

Comment (subutai) [reply htmresearch/frameworks/pytorch/functions/k_winners.py]: Yes. Delete all the (old) things!

**Comment by lscheinkman** [comment]
@subutai  This refactor may break the experiments in your branches. Should I wait for you to merge your branches first I before I merge this PR?

**Comment by subutai** [comment]
No go ahead and merge - my branch needs your changes to make progress anyway. I can update my branch afterwards.
